File: library/src/main/java/me/proxer/library/api/ProxerResponseCallAdapterFactory.java
package me.proxer.library.api;

import com.squareup.moshi.Types;
import retrofit2.Call;
import retrofit2.CallAdapter;
import retrofit2.Retrofit;

import java.lang.annotation.Annotation;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

/**
 * @author Ruben Gees
 */
final class ProxerResponseCallAdapterFactory extends CallAdapter.Factory {

    @Override
    public CallAdapter<?, ?> get(final Type returnType, final Annotation[] annotations, final Retrofit retrofit) {
        if (getRawType(returnType) != ProxerCall.class) {
            return null;
        }

        if (!(returnType instanceof ParameterizedType)) {
            throw new IllegalArgumentException("ProxerCall return type must be parameterized as "
                    + "ProxerCall<Foo> or ProxerCall<? extends Foo>");
        }

        return new ProxerResponseCallAdapter<>(getParameterUpperBound(0, (ParameterizedType) returnType));
    }

    private static final class ProxerResponseCallAdapter<R> implements CallAdapter<ProxerResponse<R>, ProxerCall<R>> {

        private final Type responseType;

        ProxerResponseCallAdapter(final Type responseType) {
            this.responseType = Types.newParameterizedType(ProxerResponse.class, responseType);
        }

        @Override
        public Type responseType() {
            return responseType;
        }

        @Override
        public ProxerCall<R> adapt(final Call<ProxerResponse<R>> call) {
            return new ProxerCall<>(call);
        }
    }
}

File: library/src/main/kotlin/me/proxer/library/internal/adapter/ProxerResponseCallAdapterFactory.kt
package me.proxer.library.internal.adapter

import com.squareup.moshi.Types
import me.proxer.library.ProxerCall
import me.proxer.library.internal.ProxerResponse
import retrofit2.Call
import retrofit2.CallAdapter
import retrofit2.Retrofit
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

/**
 * @author Ruben Gees
 */
internal class ProxerResponseCallAdapterFactory : CallAdapter.Factory() {

    override fun get(returnType: Type, annotations: Array<Annotation>, retrofit: Retrofit): CallAdapter<*, *>? {
        if (CallAdapter.Factory.getRawType(returnType) != ProxerCall::class.java) {
            return null
        }

        if (returnType !is ParameterizedType) {
            throw IllegalArgumentException(
                "ProxerCall return type must be parameterized as ProxerCall<Foo> or ProxerCall<? extends Foo>"
            )
        }

        return ProxerResponseCallAdapter<Any>(
            getParameterUpperBound(0, returnType)
        )
    }

    private class ProxerResponseCallAdapter<R> internal constructor(
        responseType: Type
    ) : CallAdapter<ProxerResponse<R>, ProxerCall<R>> {

        private val responseType = Types.newParameterizedType(ProxerResponse::class.java, responseType)

        override fun responseType(): Type {
            return responseType
        }

        override fun adapt(call: Call<ProxerResponse<R>>): ProxerCall<R> {
            return ProxerCall(call)
        }
    }
}
